proposals/generic-ref-struct.md

```diff
@@ -0,0 +1,424 @@
+
+// Can also use normal type arguments
+Swap<byte>(ref span1[0], ref span1[1]);
+```
+
+At the call site this relaxes the set of legal type arguments to inclued `ref struct` *in addition* to the types already allowed. Even though this appears in the constraint position it serves more of an anti-constraint because it expands the set of possible types. 
+
+As an expansion though it means that `where T : ref struct` is not compatible with type parameters lacking the same annotation:
+
+```c#
+void M1<U>() { }
+void M2<T>()
+    where T : ref struct
+{
+    M1<T>(); // Error: type parameter T is not compatible with U because it does not 
+             // allow `ref sturct`
+}
+```
+
+At the implementation all type parameters having the `where T : ref struct` annotation are treated as if the type is a `ref struct`. This means the type parameter has all of the limitations that would be expected a `ref struct` type. For example:
+
+- Instances of `T` cannot be boxed or appear on the heap
+- Instances of `T` can only appear as fields inside a `ref struct`
+- Instances of `T` are considered `ref struct` for the purposes of the [span safety rules](https://github.com/dotnet/csharplang/blob/master/proposals/csharp-7.2/span-safety.md) 
+- `T` cannot be used as the element of an array
+- Members on `object` such as `ToString` cannot be invoked as that implicitly boxes
+- etc ...
+
+This also means that `T : where ref struct` cannot be used where `ref struct` would be used in type positions: 
+
+```c#
+class C<T> where T : ref struct 
+{
+    // Error: field cannot be a `ref struct` inside a class
+    T field; 
+
+    // Error: array elements cannot be of type `ref struct`
+    void M(T[] array)
+    {
+
+    }
+}
+```
+
+Positionally in syntax the `ref struct` annotation must come after the type kind, such as `struct`, but before the interface list or `new()` constraint (interfaces are covered [later in the proposal](#interfaces)). The combination with type kinds has the following effects:
+
+- `where T : struct, ref struct`: describes a type parameter that can be a `struct` or `ref struct` but not other types like `class` or `interface`.
+- `where T : unmanaged, ref struct`: similar to `struct, ref struct` except that the type also meets the `unmanaged` constraint. This is believed to have limited use though as `ref struct` containing any type of `ref` data will not meet the `unmanaged` contraint and hence are likely better defined as simply `struct`.
+- `where T : notnull, ref struct`: this is the same as the `notnull` constraint but also allows for `ref struct` types to satisfy it.
+- `where T : class, ref struct`: is an error as the `ref struct` portion is meaningless when combined with `class`. 
+- `where T : default, ref struct`: is an error as the declarations are contradictory.
+- `where T : ref struct, BaseType` where `BaseType` denotes a reference type. This is an error because the `ref sruct` portion is meaningless in this context.
```

Comment:
```suggestion
- `where T : ref struct, BaseType` where `BaseType` denotes a reference type. This is an error because the `ref struct` portion is meaningless in this context.
```